source/aws_s3/s3.go

```diff
@@ -21,37 +21,72 @@ func init() {
 
 type s3Driver struct {
 	s3client   s3iface.S3API
-	bucket     string
-	prefix     string
+	config     *Config
 	migrations *source.Migrations
 }
 
+type Config struct {
+	Bucket string
+	Prefix string
+}
+
 func (s *s3Driver) Open(folder string) (source.Driver, error) {
-	u, err := url.Parse(folder)
+	driver, err := newS3Driver(folder)
 	if err != nil {
 		return nil, err
 	}
-	sess, err := session.NewSession()
-	if err != nil {
+
+	if err := driver.loadMigrations(); err != nil {
 		return nil, err
 	}
-	driver := s3Driver{
-		bucket:     u.Host,
-		prefix:     strings.Trim(u.Path, "/") + "/",
-		s3client:   s3.New(sess),
+
+	return driver, nil
+}
+
+func WithInstance(s3client s3iface.S3API, config *Config) (source.Driver, error) {
+	driver := &s3Driver{
+		config:     config,
+		s3client:   s3client,
 		migrations: source.NewMigrations(),
 	}
-	err = driver.loadMigrations()
+
+	if err := driver.loadMigrations(); err != nil {
+		return nil, err
+	}
+
+	return driver, nil
+}
+
+func newS3Driver(folder string) (*s3Driver, error) {
```

Comment:
I don't see any value in this method anymore. You can move the path parsing logic to `Open()`

Reply:
Moving it into its own function allows us to unit test the parsing logic. See the function `TestNewS3Driver` in `s3_test.go`. Moving it back to `Open()` prevents us from unit testing it as the `Open()` function will attempt to connect to a real AWS S3 bucket. Which is probably why the `Open()` function never had a unit test.

I can do it if you wish, just wanted to highlight why I moved out the logic into its own private function.

Reply:
Ok, moved most of the logic back to the `Open()` function except for the path parsing logic which is still in its own function so it can be unit tested.